lib/compat/wordpress-6.2/get-global-styles-and-settings.php

```diff
@@ -51,9 +54,40 @@ function wp_theme_has_theme_json() {
 if ( ! function_exists( 'wp_theme_has_theme_json_clean_cache' ) ) {
 	/**
 	 * Function to clean the cache used by wp_theme_has_theme_json method.
+	 *
+	 * @param string $stylesheet Directory name for the theme. Optional. Defaults to current theme.
+	 */
+	function wp_theme_has_theme_json_clean_cache( $stylesheet = '' ) {
+		if ( empty( $stylesheet ) ) {
+			$stylesheet = get_stylesheet();
+		}
+		$cache_key = sprintf( 'wp_theme_has_theme_json_%s', $stylesheet );
+		wp_cache_delete( $cache_key, 'theme_json' );
+	}
+}
+
+if ( ! function_exists( '_wp_theme_has_theme_json_clean_cache_switch_theme' ) ) {
+	/**
+	 * Clean new and old themes on switch.
+	 *
+	 * @param string  $new_name Name of the new theme.
+	 * @param WP_Theme $new_theme WP_Theme instance of the new theme.
+	 * @param WP_Theme $old_theme WP_Theme instance of the old theme.
+	 */
+	function _wp_theme_has_theme_json_clean_cache_switch_theme( $new_name, $new_theme, $old_theme ) {
+		wp_theme_has_theme_json_clean_cache( $new_theme->get_stylesheet() );
+		wp_theme_has_theme_json_clean_cache( $old_theme->get_stylesheet() );
+	}
+}
+
+if ( ! function_exists( '_wp_theme_has_theme_json_clean_cache_start_previewing_theme' ) ) {
+	/**
+	 * Clear on preview.
+	 *
+	 * @param WP_Customize_Manager $manager WP_Customize_Manager instance.
 	 */
-	function wp_theme_has_theme_json_clean_cache() {
-		wp_cache_delete( 'wp_theme_has_theme_json', 'theme_json' );
+	function _wp_theme_has_theme_json_clean_cache_start_previewing_theme( $manager ){
+		wp_theme_has_theme_json_clean_cache( $manager->get_stylesheet() );
```

Comment:
See https://github.com/WordPress/gutenberg/pull/45955/files#r1029244275 we need to clear the cache of the _previous_ theme, not the actual one.